imap_processing/utils.py

```diff
@@ -228,29 +228,77 @@ def create_dataset(
+    Convert a packet file to xarray datasets.
+
+    The packet file can contain multiple apids and these will be separated
+    into distinct datasets, one per apid. The datasets will contain the
+    ``derived_value``s of the data fields, and the ``raw_value``s if no
+    ``derived_value`` is available. If there are conversions in the XTCE
+    packet definition, the ``derived_value`` will be the converted value.
+    The dimension of the dataset will be the time field in J2000 nanoseconds.
 
     Parameters
     ----------
-    dataset : xr.Dataset
-        Dataset to update.
+    packet_file : str
+        Path to data packet path with filename.
+    xtce_packet_definition : str
+        Path to XTCE file with filename.
+    use_derived_value : bool, default True
+        Whether or not to use the derived value from the XTCE definition.
 
     Returns
     -------
-    dataset : xr.Dataset
-        Dataset with updated epoch dimension from int to datetime object.
+    datasets : dict
+        Mapping from apid to xarray dataset, one dataset per apid.
     """
-    # convert epoch to datetime
-    epoch_converted_time = met_to_j2000ns(dataset["epoch"])
-    # add attrs back to epoch
-    epoch = xr.DataArray(
-        epoch_converted_time,
-        name="epoch",
-        dims=["epoch"],
-        attrs=ConstantCoordinates.EPOCH,
-    )
-    dataset = dataset.assign_coords(epoch=epoch)
+    # Set up containers to store our data
+    # We are getting a packet file that may contain multiple apids
+    # Each apid has consistent data fields, so we want to create a
+    # dataset per apid.
+    # {apid1: dataset1, apid2: dataset2, ...}
+    data_dict: dict[int, dict] = dict()
+
+    # Set up the parser from the input packet definition
+    packet_definition = xtcedef.XtcePacketDefinition(xtce_packet_definition)
+    packet_parser = parser.PacketParser(packet_definition)
+
+    with open(packet_file, "rb") as binary_data:
+        packet_generator = packet_parser.generator(binary_data)
+        for packet in packet_generator:
+            apid = packet.header["PKT_APID"].raw_value
+            if apid not in data_dict:
+                # This is the first packet for this APID
+                data_dict[apid] = collections.defaultdict(list)
+
+            # TODO: Do we want to give an option to remove the header content?
+            packet_content = packet.data | packet.header
+
+            for key, value in packet_content.items():
+                val = value.raw_value
+                if use_derived_value:
+                    # Use the derived value if it exists, otherwise use the raw value
+                    val = value.derived_value or val
+                data_dict[apid][key].append(val)
+
+    dataset_by_apid = {}
+    # Convert each apid's data to an xarray dataset
+    for apid, data in data_dict.items():
+        # The time key is always the first key in the data dictionary on IMAP
+        time_key = next(iter(data.keys()))
```

Comment:
I am little confused about the comment. Do you mean that the time key is always the first key if we don't include the packet header and only include the data itself? That is true if that's the case. But if we include packet header than it won't the first key, right? Just want to make sure I understood your comment

Reply:
I agree it is different depending on the header, so I avoided that situation here by making it the first key right above this by appending the packet header after the data. So regardless if we want to use the header or not, the SHCOARSE will be the first entry in this content dict.
```python
packet_content = packet.data | packet.header
```

Reply:
I see. Nice! I totally missed the order in which you added those. That makes sense now. Thank you!